File: packages/testing/src/surface/schema-complexity.ts
/**
 * @fileoverview Schema complexity scoring for command input schemas.
 *
 * Scores how complex a JSON Schema is for an agent to satisfy correctly.
 * Higher scores indicate schemas that are more likely to cause agent input errors.
 */

import type { JsonSchema } from '@lushly-dev/afd-core';
import type { ComplexityBreakdown, ComplexityResult } from './types.js';

// ═══════════════════════════════════════════════════════════════════════════════
// INTERNAL TYPES
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Extends core `JsonSchema` with `type` optional (for composition-only nodes
 * like bare `oneOf`/`anyOf` wrappers) and recursive self-references.
 */
interface JsonSchemaNode
	extends Omit<JsonSchema, 'type' | 'required' | 'properties' | 'items' | 'oneOf' | 'anyOf' | 'allOf' | 'not'> {
	type?: string;
	required?: string[];
	properties?: Record<string, JsonSchemaNode>;
	items?: JsonSchemaNode;
	oneOf?: JsonSchemaNode[];
	anyOf?: JsonSchemaNode[];
	allOf?: JsonSchemaNode[];
	not?: JsonSchemaNode;
}

// ═══════════════════════════════════════════════════════════════════════════════
// HELPERS
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Detect `anyOf: [T, { type: 'null' }]` patterns from `z.nullable()`.
 * These should not count as unions — they're just nullable wrappers.
 */
function isNullableWrapper(variants: JsonSchemaNode[]): boolean {
	if (variants.length !== 2) return false;
	const hasNull = variants.some(
		(v) => v.type === 'null' || (typeof v === 'object' && 'type' in v && v.type === 'null')
	);
	return hasNull;
}

// ═══════════════════════════════════════════════════════════════════════════════
// WALKER STATE
// ═══════════════════════════════════════════════════════════════════════════════

interface WalkerState {
	fieldNames: Set<string>;
	maxDepth: number;
	unions: number;
	intersections: number;
	enums: number;
	patterns: number;
	bounds: number;
	optionalCount: number;
	totalFieldCount: number;
}

function createState(): WalkerState {
	return {
		fieldNames: new Set(),
		maxDepth: 0,
		unions: 0,
		intersections: 0,
		enums: 0,
		patterns: 0,
		bounds: 0,
		optionalCount: 0,
		totalFieldCount: 0,
	};
}

// ═══════════════════════════════════════════════════════════════════════════════
// RECURSIVE WALKER
// ═══════════════════════════════════════════════════════════════════════════════

function walk(node: JsonSchemaNode, state: WalkerState, depth: number): void {
	// Union: oneOf / anyOf
	if (node.oneOf) {
		if (!isNullableWrapper(node.oneOf)) {
			state.unions++;
		}
		for (const variant of node.oneOf) {
			walk(variant, state, depth);
		}
	}

	if (node.anyOf) {
		if (!isNullableWrapper(node.anyOf)) {
			state.unions++;
		} else {
			// Nullable wrapper — walk the non-null variant only
			for (const variant of node.anyOf) {
				if (variant.type !== 'null') {
					walk(variant, state, depth);
				}
			}
			return;
		}
		for (const variant of node.anyOf) {
			walk(variant, state, depth);
		}
	}

	// Intersection: allOf
	if (node.allOf) {
		state.intersections++;
		for (const variant of node.allOf) {
			walk(variant, state, depth);
		}
	}

	// Object
	if (node.type === 'object' && node.properties) {
		const objectDepth = depth + 1;
		if (objectDepth > state.maxDepth) {
			state.maxDepth = objectDepth;
		}

		const requiredSet = new Set(node.required ?? []);
		for (const [name, propSchema] of Object.entries(node.properties)) {
			state.fieldNames.add(name);
			state.totalFieldCount++;
			if (!requiredSet.has(name)) {
				state.optionalCount++;
			}
			walk(propSchema, state, objectDepth);
		}
		return;
	}

	// Array: transparent wrapper for depth, but recurse into items
	if (node.type === 'array' && node.items) {
		walk(node.items, state, depth);
	}

	// Constraints
	if (node.enum) {
		state.enums++;
	}
	// `const` from z.literal() is NOT counted as enum — it's a fixed value, not an agent decision point

	if (node.pattern) {
		state.patterns++;
	}
	if (node.format) {
		state.patterns++;
	}

	if (node.minimum !== undefined) state.bounds++;
	if (node.maximum !== undefined) state.bounds++;
	if (node.exclusiveMinimum !== undefined) state.bounds++;
	if (node.exclusiveMaximum !== undefined) state.bounds++;
	if (node.minLength !== undefined) state.bounds++;
	if (node.maxLength !== undefined) state.bounds++;
	if (node.minItems !== undefined) state.bounds++;
	if (node.maxItems !== undefined) state.bounds++;
}

// ═══════════════════════════════════════════════════════════════════════════════
// SCORING
// ═══════════════════════════════════════════════════════════════════════════════

/** Weight multipliers for each complexity dimension */
const WEIGHTS = {
	fields: 1,
	depth: 3,
	unions: 5,
	intersections: 2,
	enums: 1,
	patterns: 2,
	bounds: 1,
} as const;

function computeScore(breakdown: ComplexityBreakdown): number {
	return (
		breakdown.fields * WEIGHTS.fields +
		breakdown.depth * WEIGHTS.depth +
		breakdown.unions * WEIGHTS.unions +
		breakdown.intersections * WEIGHTS.intersections +
		breakdown.enums * WEIGHTS.enums +
		breakdown.patterns * WEIGHTS.patterns +
		breakdown.bounds * WEIGHTS.bounds +
		breakdown.optionalRatio
	);
}

function scoreTier(score: number): ComplexityResult['tier'] {
	if (score <= 5) return 'low';
	if (score <= 12) return 'medium';
	if (score <= 20) return 'high';
	return 'critical';
}

// ═══════════════════════════════════════════════════════════════════════════════
// PUBLIC API
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Compute the complexity score for a JSON Schema.
 *
 * Accepts `Record<string, unknown>` and casts internally to handle
 * `oneOf`/`anyOf`/`allOf` fields that exist at runtime but aren't
 * in the core `JsonSchema` type.
 */
export function computeComplexity(schema: Record<string, unknown>): ComplexityResult {
	const state = createState();
	walk(schema as unknown as JsonSchemaNode, state, 0);

	const optionalRatio =
		state.totalFieldCount > 0 ? Math.floor((state.optionalCount / state.totalFieldCount) * 4) : 0;

	const breakdown: ComplexityBreakdown = {
		fields: state.fieldNames.size,
		depth: state.maxDepth,
		unions: state.unions,
		intersections: state.intersections,
		enums: state.enums,
		patterns: state.patterns,
		bounds: state.bounds,
		optionalRatio,
	};

	const score = computeScore(breakdown);

	return {
		score,
		tier: scoreTier(score),
		breakdown,
	};
}

File: python/src/afd/testing/surface/schema_complexity.py
"""
Schema complexity scoring for command input schemas.

Scores how complex a JSON Schema is for an agent to satisfy correctly.
Higher scores indicate schemas that are more likely to cause agent input errors.

Port of packages/testing/src/surface/schema-complexity.ts
"""

from __future__ import annotations

import math
from typing import Any

from afd.testing.surface.types import ComplexityBreakdown, ComplexityResult

# Weight multipliers for each complexity dimension
WEIGHTS = {
	"fields": 1,
	"depth": 3,
	"unions": 5,
	"intersections": 2,
	"enums": 1,
	"patterns": 2,
	"bounds": 1,
}


def _is_nullable_wrapper(variants: list[dict[str, Any]]) -> bool:
	"""Detect anyOf: [T, {type: 'null'}] patterns from nullable wrappers."""
	if len(variants) != 2:
		return False
	return any(v.get("type") == "null" for v in variants)


def _walk(node: dict[str, Any], state: dict[str, Any], depth: int) -> None:
	"""Recursively walk a JSON Schema node and accumulate state."""
	# Union: oneOf / anyOf
	if "oneOf" in node:
		one_of = node["oneOf"]
		if isinstance(one_of, list):
			if not _is_nullable_wrapper(one_of):
				state["unions"] += 1
			for variant in one_of:
				if isinstance(variant, dict):
					_walk(variant, state, depth)

	if "anyOf" in node:
		any_of = node["anyOf"]
		if isinstance(any_of, list):
			if not _is_nullable_wrapper(any_of):
				state["unions"] += 1
				for variant in any_of:
					if isinstance(variant, dict):
						_walk(variant, state, depth)
			else:
				# Nullable wrapper - walk the non-null variant only
				for variant in any_of:
					if isinstance(variant, dict) and variant.get("type") != "null":
						_walk(variant, state, depth)
				return

	# Intersection: allOf
	if "allOf" in node:
		all_of = node["allOf"]
		if isinstance(all_of, list):
			state["intersections"] += 1
			for variant in all_of:
				if isinstance(variant, dict):
					_walk(variant, state, depth)

	# Object
	if node.get("type") == "object" and "properties" in node:
		object_depth = depth + 1
		if object_depth > state["max_depth"]:
			state["max_depth"] = object_depth

		required_set = set(node.get("required", []))
		properties = node.get("properties", {})
		if isinstance(properties, dict):
			for name, prop_schema in properties.items():
				state["field_names"].add(name)
				state["total_field_count"] += 1
				if name not in required_set:
					state["optional_count"] += 1
				if isinstance(prop_schema, dict):
					_walk(prop_schema, state, object_depth)
		return

	# Array
	if node.get("type") == "array" and "items" in node:
		items = node["items"]
		if isinstance(items, dict):
			_walk(items, state, depth)

	# Constraints
	if "enum" in node:
		state["enums"] += 1

	if "pattern" in node:
		state["patterns"] += 1
	if "format" in node:
		state["patterns"] += 1

	for bound_key in (
		"minimum", "maximum", "exclusiveMinimum", "exclusiveMaximum",
		"minLength", "maxLength", "minItems", "maxItems",
	):
		if bound_key in node:
			state["bounds"] += 1


def _compute_score(breakdown: ComplexityBreakdown) -> float:
	"""Compute weighted complexity score."""
	return (
		breakdown.fields * WEIGHTS["fields"]
		+ breakdown.depth * WEIGHTS["depth"]
		+ breakdown.unions * WEIGHTS["unions"]
		+ breakdown.intersections * WEIGHTS["intersections"]
		+ breakdown.enums * WEIGHTS["enums"]
		+ breakdown.patterns * WEIGHTS["patterns"]
		+ breakdown.bounds * WEIGHTS["bounds"]
		+ breakdown.optional_ratio
	)


def _score_tier(score: float) -> str:
	"""Determine complexity tier from score."""
	if score <= 5:
		return "low"
	if score <= 12:
		return "medium"
	if score <= 20:
		return "high"
	return "critical"


def compute_complexity(schema: dict[str, Any]) -> ComplexityResult:
	"""Compute the complexity score for a JSON Schema."""
	state: dict[str, Any] = {
		"field_names": set(),
		"max_depth": 0,
		"unions": 0,
		"intersections": 0,
		"enums": 0,
		"patterns": 0,
		"bounds": 0,
		"optional_count": 0,
		"total_field_count": 0,
	}

	_walk(schema, state, 0)

	optional_ratio = (
		math.floor((state["optional_count"] / state["total_field_count"]) * 4)
		if state["total_field_count"] > 0
		else 0.0
	)

	breakdown = ComplexityBreakdown(
		fields=len(state["field_names"]),
		depth=state["max_depth"],
		unions=state["unions"],
		intersections=state["intersections"],
		enums=state["enums"],
		patterns=state["patterns"],
		bounds=state["bounds"],
		optional_ratio=optional_ratio,
	)

	score = _compute_score(breakdown)
	tier = _score_tier(score)

	return ComplexityResult(score=score, tier=tier, breakdown=breakdown)
